### ISSUE TITLE
Feature Request: Support E2E tests written in TypeScript

### ISSUE THREAD

**Issue by joeskeen** [NativeScript/nativescript-dev-appium#13]
For apps written in Angular 2 with TypeScript, it would be nice to be able to author my E2E tests using TypeScript as well.  This is possible today, but the workflow could be improved.  Here are the current options:  

1. Run `tsc` on your tests before running them.  This is less than ideal due to the manual compilation step and the extra work required to exclude your JS files from GIT, etc.
1. Mocha does support TypeScript (or any other transpiler) via command-line arguments such as this:
    ```
    --compilers ts:ts-node/register
    ```
    
    The way that the code is written, the Mocha options passed in are not configurable (see https://github.com/NativeScript/nativescript-dev-appium/blob/master/test-runner.js#L62 and https://github.com/NativeScript/nativescript-dev-appium/blob/master/test-runner.js#L86)
    
    Fortunately, Mocha itself provides a solution: a `mocha.opts` file, where each line is a command-line argument to pass into the Mocha process.  Unfortunately, though, the `mocha.opts` file must be found at `./test/mocha.opts`, unless you pass a `--opts ./e2e-tests/mocha.opts` parameter, so you either supply that option every time or have both a `./test` folder for Mocha configuration and a `./e2e-tests` folder for your actual tests (not to mention `./app/tests` for your unit tests).  

So, how would be the best way to solve this?  Here are my ideas (feel free to come up with better ones):

* Add a `--typescript` flag that sets up the `--compilers` option.  This would be a low learning curve for newbies to get their TS tests up and running with minimal configuration, but would be just the first of many transpiler flags that would have to be required (Babel, CoffeeScript, [COBOL](https://github.com/ajlopez/CobolScript), I mean, who knows...). Also, this option would make it harder to target a specific version of TypeScript, etc.
* Add a `--mochaOpts` flag that will concatenate the specified args with the `mochaOpts` array before calling Mocha.  This would allow for more flexibility, but would force a more lengthy command for each test run (although aliasing would make this less of an issue)
* Add support for a `mocha.opts` file in the `e2e-tests` folder. (I like this option the best)  The workflow could be thus:
    ```
    var mochaOptsFile = path.join(projectDir, 'e2e-tests', 'mocha.opts');
    if (fs.existsSync(mochaOptsFile)) {
      mochaOpts.push('--opts');
      mochaOpts.push(mochaOptsFile);
    }
    ```

Adding support for custom Mocha options would open up many more possibilities, such as custom reporters or longer timeouts.  Again, all this is possible today, but not straightforward.

What do you think @hdeshev?  Would you be open to a PR with the suggested code above?

**Comment by joeskeen** [comment]
Along with this, it would also be nice to have a TypeScript declaration file for 'nativescript-dev-appium' I would be happy to throw one together.

Improving TS support for this is important to me as it will help encourage those using TypeScript to better test their applications.

**Comment by hdeshev** [comment]
This sounds great. The only thing that's prevented me from doing it in the past is that it's a bit unclear which test template we should generate when `npm install`-ing the package: JavaScript or TypeScript. Perhaps we could generate both sample tests and let the user delete the one(s) s/he doesn't need.

**Comment by joeskeen** [comment]
I know when you invoke `tns test init`, you get a choice of testing framework you would like to use.  Perhaps we could do something similar for the end-to-end tests? I know that would be more work but could be beneficial long-term.

I also hit a snag in the type definitions for my end-to-end tests.  I can describe the top-level of the API easy enough, but when you start chaining Appium stuff to expectations and back again, it seems that we lose the Appium commands from the autocompletion.  I will need a little time to work out the best way to get around this (as the chain at that point is an object that contains all webdriver commands plus it's also a promise plus it's also a chai-as-promised expectation).

As far as supporting the `mocha.opts` file in `./e2e-tests`, though, that is something IMHO that should be done sooner than later to open up custom reporters and such, I will submit that PR first while I'm mulling over how to get first-class TS support working.

**Comment by hdeshev** [comment]
We have some typings in the nativescript-cli project that I usually steal whenever I need to add tests. I hope they help you too:

- https://github.com/telerik/mobile-cli-lib/blob/a650ccc520ae46655bfb9e08160c676d39814d00/test/definitions/mocha.d.ts
- https://github.com/telerik/mobile-cli-lib/blob/a650ccc520ae46655bfb9e08160c676d39814d00/test/definitions/chai.d.ts

**Comment by joeskeen** [comment]
I would usually just run `npm install --save-dev @types/mocha @types/chai-as-promised` to get the latest typings from DefinitelyTyped; however I've come across some use cases where the typings get in the way a bit, since all the commands are chained together and after an assertion the inferred type is no longer including the webdriver commands, as in this example that runs just fine:

```ts
return driver
        .waitForElementByAccessibilityId("myId", 300000) //wd command
        .elementByAccessibilityId("myId") //wd command
        .should.eventually.exist //assertions
        .text() //wd command, this line will cause a TS error
        .should.eventually.match(/^Hello/); //assertions
```

For cases like these, it may be necessary to modify the typings file to make them return a `this` type or something similar to keep the types flowing through the test case without sacrificing the type checking completely by just casting the whole thing to `<any>`.

I've done a lot of work with type definitions in the past, so I'm confident I can get to the bottom of this, but it may take a little time as I'm a little busy at present.

**Comment by SvetoslavTsenov** [comment]
Hi, @joeskeen, currently we are working on that and you can give a try to our @next version. Any feedback will be highly appreciated. We are about to release a new official version that will support also an image comparison and a bunch of other features.

**Comment by SvetoslavTsenov** [comment]
This feature is already available with nativescript-dev-appium 3.0.0 and above
